### PR TITLE
Implement generic `MultiControlledBloq`  using single-controlled subbloq + And ladder.

### PR DESCRIPTION
**Pull request by anurudhp**
#1131 

- Add `ControlledViaAnd` that computes a single qubit control using `CtrlSpecAnd`, and uses a single-controlled subbloq.
- Refactor `MultiControlPauli` to delegate to this. For the moment I did not remove it fully as it breaks too many things that switch between cirq-style and qualtran-style defs. This should be cleaned up in a follow-up PR.

### CODE REVIEW

File: qualtran/bloqs/mcmt/multi_controlled_bloq.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+from qualtran.bloqs.basic_gates import XGate
+from qualtran.bloqs.mcmt.ctrl_spec_and import CtrlSpecAnd
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, Signature, SoquetT
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class MultiControlledBloq(Bloq):
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/multi_controlled_bloq.py:30]: I see @mpharrigan's point that having both `Controlled(sub_bloq, ctrl_spec)` and `MultiControlledBloq(sub_bloq, ctrl_spec)` can be confusing and the implementation of both would actually be identical. 

Regarding interface vs implementation, it may make things cleaner if we rename `MultiControlledBloq` to be more descriptive of the implementation, like `ControlledViaAndLadder()` and explain that the implementation performs `CtrlSpecAnd` -> `Bloq.controlled()` -> `CtrlSpecAnd.adjoint()` to decompose a bloq controlled by arbitrary `CtrlSpec` to a singly controlled Bloq via an And / And^{dagger} ladder and uses $num_controls$ ancilla qubits. 

Then, this specific implementation can be used by users who want to depend upon this behavior instead of the behavior getting potentially updated if we decide to update the decomposition of `Controlled` Bloq later. 

`MultiControlledPauli` can get removed and we will simply update the call sites to use `ControlledViaAndLadder(ctrl_spec, XGate())`

Comment (anurudhp) [reply qualtran/bloqs/mcmt/multi_controlled_bloq.py:30]: I like the bloq name `ControlledViaAndLadder`, will use that

File: qualtran/bloqs/mcmt/multi_control_multi_target_pauli.py [hunk 2]
```diff
@@ -141,30 +143,34 @@ def concrete_cvs(self) -> Tuple[int, ...]:
...
+        return ControlledViaAnd(self.target_bloq, ctrl_spec)
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        if is_symbolic(self.cvs):
+            raise DecomposeTypeError(f"cannot decompose {self} with symbolic {self.cvs=}")
+
+        target = soqs.pop('target')
+        (target_reg_name,) = [reg.name for reg in self.target_bloq.signature]
+        if self.n_ctrls == 0:
+            # TODO discuss if we should remove support for this case.
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/multi_control_multi_target_pauli.py:164]: This was specifically added after https://github.com/quantumlib/Qualtran/issues/677

File: qualtran/bloqs/mcmt/ctrl_spec_and.py [hunk 3]
```diff
@@ -224,3 +226,111 @@ def _ctrl_on_multiple_values() -> CtrlSpecAnd:
     bloq_cls=CtrlSpecAnd,
     examples=(_ctrl_on_int, _ctrl_on_bits, _ctrl_on_nd_bits, _ctrl_on_multiple_values),
 )
+
+
+@frozen
+class ControlledViaAnd(Bloq):
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/ctrl_spec_and.py:232]: Let's add it into a new file. It's good to have 1 bloq per file

Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/ctrl_spec_and.py:232]: Also, please add tests for this class.

File: qualtran/bloqs/mcmt/ctrl_spec_and.py [hunk 4]
```diff
@@ -224,3 +226,111 @@ def _ctrl_on_multiple_values() -> CtrlSpecAnd:
...
+            soqs |= {
+                name: ctrl_soqs[inner_ctrl_reg.name]
+                for name, inner_ctrl_reg in zip(
+                    self.ctrl_reg_names, ctrl_spec_and_bloq.control_registers
+                )
+            }
+
+        return soqs
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> set['BloqCountT']:
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/ctrl_spec_and.py:326]: Please override `wire_symbol` to enable nice diagrams. Override `__str__` to display pretty strings. Preferably override `adjoint` such that bloq simply delegates to `ControlledViaAnd(self.subbloq.adjoint(), self.ctrl_spec)`

File: qualtran/bloqs/mcmt/ctrl_spec_and.py [hunk 5]
```diff
@@ -224,3 +226,111 @@ def _ctrl_on_multiple_values() -> CtrlSpecAnd:
     bloq_cls=CtrlSpecAnd,
     examples=(_ctrl_on_int, _ctrl_on_bits, _ctrl_on_nd_bits, _ctrl_on_multiple_values),
 )
+
+
+@frozen
+class ControlledViaAnd(Bloq):
+    """Reduces a generic control bloq to a singly-controlled bloq using an And ladder.
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/ctrl_spec_and.py:233]:
```suggestion
    """Reduces a generic controlled bloq to a singly-controlled bloq using an And ladder.
```

File: qualtran/bloqs/mcmt/ctrl_spec_and.py (hunk 3 above)
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/ctrl_spec_and.py:232]: @mpharrigan Do you think the Bloq should derive from `Controlled` instead of `Bloq` so we can reuse some of the infrastructure like `wire_symbol`, `classical_action`, getting `ctrl_reg_names` and `ctrl_regs` ?

File: qualtran/bloqs/mcmt/ctrl_spec_and.py [hunk 6]
```diff
@@ -224,3 +226,111 @@ def _ctrl_on_multiple_values() -> CtrlSpecAnd:
...
+        return self._signature_and_ctrl_regs_from_controlled[0]
+
+    @cached_property
+    def ctrl_reg_names(self) -> tuple[str, ...]:
+        return self._signature_and_ctrl_regs_from_controlled[1]
+
+    @cached_property
+    def _signature_and_ctrl_regs_from_controlled(self) -> tuple[Signature, tuple[str, ...]]:
+        cbloq = Controlled(self.subbloq, self.ctrl_spec)
+        return cbloq.signature, tuple(cbloq.ctrl_reg_names)
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/ctrl_spec_and.py:260]: I think it's worth pulling out the logic in `Controlled.ctrl_reg_names` and `Controlled.ctrl_regs` into a free function `def ctrl_regs_from_ctrl_spec(ctrl_spec: CtrlSpec, subbloq: Optional[Bloq] = None)` which will return you a list of ctrl registers given a ctrl spec and subbloq. Then we won't need to allocate the `Controlled` subbloq only to define the signature of this bloq.

**Review by tanujkhattar — COMMENTED** [review]
Looking good. Left a final round of comments

File: qualtran/bloqs/mcmt/ctrl_spec_and.py (hunk 3 above)
Comment (anurudhp) [reply qualtran/bloqs/mcmt/ctrl_spec_and.py:232]: Oh I didn't think of this, I really like the idea! If it doesn't break anything, I'd prefer going with this, makes the interface cleaner.

File: qualtran/bloqs/mcmt/ctrl_spec_and.py [hunk 7]
```diff
@@ -224,3 +226,111 @@ def _ctrl_on_multiple_values() -> CtrlSpecAnd:
...
+        counts: set['BloqCountT'] = {(self.subbloq.controlled(), 1)}
+
+        if self._is_single_bit_control():
+            if self._single_control_value == 0:
+                counts |= {(XGate(), 2)}
+        else:
+            ctrl = CtrlSpecAnd(self.ctrl_spec)
+            counts |= {(ctrl, 1), (ctrl.adjoint(), 1)}
+
+        return counts
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/mcmt/ctrl_spec_and.py]: ah thanks! fixed

File: qualtran/bloqs/mcmt/ctrl_spec_and.py (hunk 3 above)
Comment (anurudhp) [reply qualtran/bloqs/mcmt/ctrl_spec_and.py]: inherited from Controlled

File: qualtran/_infra/controlled.py [hunk 8]
```diff
@@ -358,15 +358,19 @@ def signature(self) -> 'Signature':
         return Signature(self.ctrl_regs + tuple(self.subbloq.signature))
 
     def decompose_bloq(self) -> 'CompositeBloq':
+        return Bloq.decompose_bloq(self)
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', **initial_soqs: 'SoquetT'
+    ) -> Dict[str, 'SoquetT']:
```
Comment (tanujkhattar) [inline qualtran/_infra/controlled.py:365]: Why this change? Seems unrelated to the PR

**Review by tanujkhattar — APPROVED** [review]
LGTM

File: qualtran/_infra/controlled.py (hunk 8 above)
Comment (tanujkhattar) [reply qualtran/_infra/controlled.py:365]: Okay nvm, got it.

File: qualtran/bloqs/mcmt/controlled_via_and.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+from qualtran.bloqs.basic_gates import XGate
+from qualtran.bloqs.mcmt.ctrl_spec_and import CtrlSpecAnd
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, SoquetT
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class ControlledViaAnd(Controlled):
```
Comment (tanujkhattar) [inline qualtran/bloqs/mcmt/controlled_via_and.py:31]: @anurudhp You still didn't add tests for this class - Please create a follow up PR which adds `qualtran/bloqs/mcmt/controlled_via_and_test.py`, adds bloq examples and ties this bloq to the docs infrastructure. Sorry if I merged it too soon.

Comment (anurudhp) [reply qualtran/bloqs/mcmt/controlled_via_and.py:31]: oops, I was about to add, but no worries! will add in a new PR
